oml/models/vit/clip.py

```diff
@@ -0,0 +1,240 @@
+        x = x.permute(1, 0, 2)  # LND -> NLD
+
+        x = self.ln_post(x[:, 0, :])
+
+        if self.proj is not None:
+            x = x @ self.proj
+
+        return x
+
+
+class ViTCLIPExtractor(IExtractor):
+    pretrained_models: Dict[str, Any] = {}  # there are too many pretrained architectures, so we will skip them
+
+    def __init__(
+        self,
+        arch: Optional[str] = None,
+        normalise_features: bool = True,
+        embed_dim: int = 512,
+        image_resolution: int = 224,
+        layers: int = 12,
+        width: int = 768,
+        patch_size: int = 32,
+        heads: int = 8,
+        weights: Optional[str] = None,
+        strict_load: bool = True,
+    ):
+        """
+        Args:
+            weights: Path to weights or ``None`` for randomly initialized model's weights.
+             Available pretrained checkpoints are currently matching with possible architectures (``arch`` param).
+            arch: Might be one of ``openai_vitb16_224``, ``openai_vitb32_224``, ``openai_vitl14_224``, ``openai_vitl14_336``.
+            normalise_features: Set ``True`` to normalise output features
+            embed_dim: Embedding dimension.
+            image_resolution: Input image resolution.
+            layers: Number of layers in ViT.
+            width: ViT's width. Default is 3 * 128.
+            patch_size: Convolutional encoder patch size.
+            heads: Number of heads in MHA.
+            strict_load: Whether the weights needed to be loaded strictly. Doesn't work with OpenAI's models.
+        """
+
+        super().__init__()
+
+        self.normalize = normalise_features
+
+        if not arch:
+            self.visual = VisionTransformer(
+                input_resolution=image_resolution,
+                patch_size=patch_size,
+                layers=layers,
+                width=width,
+                heads=heads,
+                output_dim=embed_dim,
+            )
+            if weights:
+                state_dict = torch.load(Path(weights), map_location="cpu")["state_dict"]
+                state_dict = remove_prefix_from_state_dict(state_dict, trial_key="conv1.weight")
+                self.visual.load_state_dict(state_dict=state_dict, strict=strict_load)
+        else:
+            cfg = get_vit_config_by_name(arch)
+            embed_dim = cfg["embed_dim"]
+            image_resolution = cfg["image_resolution"]
+            layers = cfg["layers"]
+            width = cfg["width"]
+            patch_size = cfg["patch_size"]
+            heads = cfg["heads"]
+            weights = cfg["weights"]
+            self.visual = torch.jit.load(Path(weights), map_location="cpu").visual
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        with torch.autocast(x.device.type):
+            if not self.normalize:
+                return self.visual.forward(x)
+            else:
+                res = self.visual.forward(x)
+                return res / res.norm(dim=1, keepdim=True)
```

Comment:
please, check how we normalise features in other models

Reply:
This normalization is directly from the original paper. I thought it is much better to explicitly keep it the same.